File: src/main/java/com/matamercer/microblog/web/api/activitypub/ActivityPubController.java
package com.matamercer.microblog.web.api.activitypub;

import com.matamercer.microblog.web.api.activitypub.render.KeyRender;
import com.matamercer.microblog.web.api.activitypub.render.PersonRender;
import com.matamercer.microblog.Exceptions.NotFoundException;
import com.matamercer.microblog.models.entities.User;
import com.matamercer.microblog.models.repositories.activitypub.UserKeyPairRepository;
import com.matamercer.microblog.models.repositories.UserRepository;
import com.matamercer.microblog.utilities.EnvironmentUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.net.UnknownHostException;
import java.util.Optional;


@RestController
@RequestMapping("/activitypub")
class ActivityPubController {

//    @RequestMapping("/secured")
//    public String secured() {
//        System.out.println("Inside secured()");
//        return "Hello user !!! : ";
//    }

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private UserKeyPairRepository userKeyPairRepository;

    @Autowired
    private EnvironmentUtil environmentUtil;

    @GetMapping("/users/{userId}")
    public PersonRender getPerson(@PathVariable("userId") Long userId) throws UnknownHostException {
        Optional<User> userOptional = userRepository.findById(userId);
        PersonRender personRender;
        if(userOptional.isPresent()){
            User user = userOptional.get();
            personRender = new PersonRender();
            personRender.setId(environmentUtil.getServerUrl() + "/activitypub/users/" + user.getId());
            personRender.setInbox(personRender.getId() + "/inbox");
            personRender.setName(user.getUsername());
            personRender.setOutbox(personRender.getId() + "/outbox");
            personRender.setPreferredUsername(user.getUsername());

            KeyRender keyRender = new KeyRender();
            keyRender.setOwner(environmentUtil.getServerUrl() + "/activitypub/users/" + user.getId());
            keyRender.setId(keyRender.getOwner() + "/main-key");
            keyRender.setPublicKeyPem(userKeyPairRepository.findByUser(user).getPublicKey());
            personRender.setPublicKey(keyRender);
        }else{
            throw new NotFoundException();
        }

        return personRender;
    }



}

File: src/main/java/com/matamercer/microblog/web/api/activitypub/ActivityPubController.kt
package com.matamercer.microblog.web.api.activitypub

import com.matamercer.microblog.exceptions.NotFoundException
import com.matamercer.microblog.models.repositories.UserKeyPairRepository
import com.matamercer.microblog.models.repositories.UserRepository
import com.matamercer.microblog.utilities.EnvironmentUtil
import com.matamercer.microblog.web.api.activitypub.render.KeyRender
import com.matamercer.microblog.web.api.activitypub.render.PersonRender
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController
import java.net.UnknownHostException

@RestController("/activitypub")
class ActivityPubController(
    private val userRepository: UserRepository? = null,
    private val userKeyPairRepository: UserKeyPairRepository? = null,
    private val environmentUtil: EnvironmentUtil? = null
) {

    @GetMapping("/users/{userId}")
    @Throws(UnknownHostException::class)
    fun getPerson(@PathVariable("userId") userId: Long): PersonRender {
        val user = userRepository?.findById(userId)?.orElseThrow {
            throw NotFoundException()
        }
        val keyRender = KeyRender(
            id = environmentUtil?.serverUrl + "/activitypub/users/" + user!!.id + "/main-key",
            owner = environmentUtil?.serverUrl + "/activitypub/users/" + user.id,
            publicKeyPem = userKeyPairRepository!!.findByUser(user)!!.publicKey
        )
        return PersonRender(
            id = environmentUtil!!.serverUrl + "/activitypub/users/" + user.id,
            inbox = environmentUtil.serverUrl + "/activitypub/users/" + user.id + "/inbox",
            name = user.username,
            preferredUsername = user.username,
            outbox = environmentUtil.serverUrl + "/activitypub/users/" + user.id + "/outbox",
            publicKey = keyRender
        )
    }
}